src/tests/cli.rs

```diff
@@ -1,88 +1,133 @@
-// Tests for the amber CLI binary itself.
-// Make sure to run `cargo build` before running these tests.
+// Tests for the amber CLI binary functionality using internal APIs.
+// These tests use internal compilation and execution functions instead of
+// relying on the external binary, making them more reliable and faster.
 
-use assert_cmd::prelude::*;
-use predicates::prelude::*;
-use std::io::Write;
-use std::process::Command;
-use tempfile::NamedTempFile;
+use crate::compiler::{AmberCompiler, CompilerOptions};
 
 // Test that the bash error code is forwarded to the exit code of amber.
 #[test]
 fn bash_error_exit_code() -> Result<(), Box<dyn std::error::Error>> {
-    let mut cmd = Command::cargo_bin(env!("CARGO_PKG_NAME"))?;
-
-    let mut file = NamedTempFile::new()?;
-
-    writeln!(
-        file,
-        r#"
-        main {{
+    let amber_code = r#"
+        main {
             $ notexistingcommand $?
-        }}
-        "#
-    )?;
-
-    // Changes locale to default to prevent locale-specific error messages.
-    cmd.env("LC_ALL", "C")
-        .arg("run")
-        .arg("--no-proc")
-        .arg("*")
-        .arg(file.path());
-
-    cmd.assert()
-        .failure()
-        .stderr(predicate::str::contains("notexistingcommand: command not found"))
-        .code(127);
-
+        }
+        "#;
+    
+    // Create compiler with no postprocessors (equivalent to --no-proc *)
+    let mut options = CompilerOptions::default();
+    options.no_proc = vec!["*".to_string()];
+    let compiler = AmberCompiler::new(amber_code.to_string(), None, options);
+    
+    // Compile the amber code to bash
+    let (messages, bash_code) = compiler.compile()
+        .map_err(|e| format!("Compilation failed: {}", e.message.unwrap_or_default()))?;
+    
+    // Show any compiler messages (shouldn't be any for this test)
+    messages.iter().for_each(|m| m.show());
+    
+    // Execute the bash code and check the exit status
+    let exit_status = AmberCompiler::execute(bash_code, vec![])?;
+    
+    // Verify that the command failed with exit code 127 (command not found)
+    assert_eq!(exit_status.code(), Some(127), "Expected exit code 127 for command not found");
+    
     Ok(())
 }
```

Comment:
This function actually requires running the compiled bash code in a subshell and we have to check if it exits with the correct exit code